tools/buildsys/src/public/variant.rs

```diff
@@ -0,0 +1,412 @@
+/// let variant = Variant::from_env().unwrap();
+///
+/// assert_eq!(variant.version(), "1.21");
+///
+/// // In a `build.rs` file, you may want to emit cfgs that you can use for conditional compilation.
+/// variant.emit_cfgs();
+/// ```
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Variant {
+    variant: String,
+    platform: String,
+    runtime: String,
+    family: String,
+    version: Option<String>,
+    variant_type: Option<String>,
+}
+
+impl Variant {
+    /// Create a new `Variant` from a dash-delimited string. The first two tuple positions,
+    /// `platform` and `runtime` are required. The next two, representing `variant_version` and
+    /// `variant_type`, are optional.
+    ///
+    /// # Valid Values
+    ///
+    /// - `aws-dev`
+    /// - `vmware-k8s-1.21`
+    /// - `aws-k8s-1.21-nvidia`
+    /// - `aws-k8s-1.21-nvidia-some-additional-ignored-tuple-positions`
+    ///
+    /// # Invalid Values
+    ///
+    /// - `aws`
+    /// - `aws-dev-`
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use buildsys::Variant;
+    /// let variant = Variant::new("aws-k8s").unwrap();
+    /// assert_eq!(variant.family(), "aws-k8s");
+    /// ```
+    pub fn new<S: Into<String>>(value: S) -> Result<Self> {
+        Self::parse(value)
+    }
+
+    /// Create a new `Variant` from the `VARIANT` environment variable's value. The environment
+    /// variable must exist and its value must be a valid variant string tuple.
+    pub fn from_env() -> Result<Self> {
+        let value = std::env::var(VARIANT_ENV).context(error::VariantEnvSnafu)?;
+        Variant::new(value)
+    }
+
+    /// The variant's platform. This is the first member of the tuple. For example, in `vmware-dev`,
+    /// `vmware` is the platform.
+    pub fn platform(&self) -> &str {
+        &self.platform
+    }
+
+    /// The variant's runtime. This is the second member of the tuple. For example, in
+    /// `metal-k8s-1.21`, `k8s` is the `runtime`.
+    pub fn runtime(&self) -> &str {
+        &self.runtime
+    }
+
+    /// The variant's family. This is the `platform` and `runtime` together. For example, in
+    /// `aws-k8s-1.21`, `aws-k8s` is the `family`.
+    pub fn family(&self) -> &str {
+        &self.family
+    }
+
+    /// The variant's version. This is the optional third value in the variant string tuple. For
+    /// example for `aws-ecs-1` the `version` is `1`. If the `version` does not exist,
+    /// [`DEFAULT_VARIANT_VERSION`] is returned.
+    pub fn version(&self) -> &str {
+        self.version.as_deref().unwrap_or(DEFAULT_VARIANT_VERSION)
+    }
+
+    /// The variant's type. This is the optional fourth value in the variant string tuple. For
+    /// example for `aws-k8s-1.21-nvidia` the `variant_type` is `nvidia`. If the `variant_type` does
+    /// not exist, [`DEFAULT_VARIANT_TYPE`] is returned.
+    pub fn variant_type(&self) -> &str {
+        self.variant_type.as_deref().unwrap_or(DEFAULT_VARIANT_TYPE)
+    }
```

Comment:
Maybe bring back `flavor` for this concept? Naming is hard, but I'd like to get away from terminology with a counterpart in genetics.